src/wasm/wasm-type.cpp

```diff
@@ -393,6 +393,14 @@ bool Type::isException() const {
 
 bool Type::isNullable() const {
   if (isBasic()) {
+    if (id == i31ref) {
+      // i31ref is always non-nullable - there is no way to declare a nullable
+      // i31 reference, unlike other reference types. For now, allow it in vars
```

Comment:
Isn't `(ref null i31)` legal?

Reply:
Good question... the binaryen parser doesn't allow it. But maybe we should, as it seems like `i31` is a heap type...?

In that case, we could just remove `i31ref` entirely, if it's just a wat shorthand for `(ref i31)`? That is, we could remove i31ref from the entire codebase, and whenever we need one, do `Type(HeapType::i31, NonNullable)`.

Reply:
Well, I wouldn't go as far as removing `i31ref` from the codebase entirely, unless we also want to remove `anyref`, `externref`, etc. for the same reasons. But we should be using `(ref null i31)` rather than `i31ref` everywhere we are currently using e.g. `externref` in place of `(ref extern)`.

Reply:
@tlively I'm not sure I follow your last sentence. What is the analogy with externref saying?

My point maybe wasn't clear before. What I am saying is this: right now we have `i31ref`, and we cannot read `(ref null i31)` (with or without `null`). If I add the ability to read that, then now we have two internal representations of the same type, `Type::i31ref` and `Type(HeapType::BasicHeapType::i31, NonNullable)`.

The latter would not count as `Type::isBasic()` right now, so we'd need to fix that. I think we'd need to fix equality checking by making sure the two are interned in the same way, but I'm not sure. Instead, it seems simpler to me to have a single internal representation (the second of the two above), and to basically turn `i31ref` into sugar in the text format, which is immediately lowered in the s-parser. Thoughts?

(That all seems like annoying work, so I'd prefer if there's a better option!)

Reply:
Due to canonicalization, `Type(HeapType::BasicHeapType::i31, NonNullable) == Type::i31ref`, just like `Type(HeapType::BasicHeapType::any, Nullable) == Type::anyref`. That means that `Type::isBasic()` returns `true` either way.

`i31ref`, `anyref`, `externref`, etc. are all just shorthands for `(ref i31)`, `(ref null any)`, `(ref null extern)`, etc. We could remove all these shorthands from the codebase, but it wouldn't make sense to remove just one of them.

I'm suggesting that we parse `i31ref` (i.e. `(ref i31)`) as `(ref null i31)` just like we currently parse `(ref any)` as `(ref null any)` (i.e. `anyref`).